File: AWS/iac-microbenchmark/ec2/python/sha256.py
#!/usr/bin/env python3
import boto3
import base64
import json
import os
import sys

# Initialize the boto3 client for KMS
kms_client = boto3.client('kms')

SIGN_ALGORITHM = 'HMAC_SHA_256'

def main():

    # Get the name argument from sys.argv
    request_json_raw = sys.argv[1]

    request_json = json.loads(request_json_raw)

    message = request_json["message"]

    sha_kms_key_id = os.environ['SHA256_KMS_KEY_ARN']

    # Convert the message to bytes
    message_bytes = message.encode('utf-8')
    # Use KMS to generate (HMAC) the message
    response = kms_client.generate_mac(
        KeyId=sha_kms_key_id,
        Message=message_bytes,
        MacAlgorithm=SIGN_ALGORITHM
    )

    # Base64 encode the signature for output
    signature = base64.b64encode(response['Mac']).decode('utf-8')
    result_dict = {
        "signature" : signature
    }

    print(json.dumps(result_dict))

if __name__ == "__main__":
    main()

File: AWS/iac-microbenchmark/ec2/typescript/sha256.ts
#!/usr/bin/env node

import { KMSClient, GenerateMacCommand } from '@aws-sdk/client-kms';
import * as process from 'process';

// Initialize the AWS KMS client
const kmsClient = new KMSClient({
  region: 'us-east-1' // Specify the desired region
});// Define the signing algorithm

const SIGN_ALGORITHM = 'HMAC_SHA_256';

async function main(): Promise<void> {
  try {
    // Get the JSON input from command line arguments
    const requestJsonRaw: string = process.argv[2];

    // Parse the JSON input
    const requestJson = JSON.parse(requestJsonRaw);

    // Extract the message from the parsed JSON
    const message: string = requestJson["message"];

    // Retrieve the KMS key ARN from environment variables
    const shaKmsKeyId = process.env['SHA256_KMS_KEY_ARN'];

    // Check if the KMS key ARN is provided
    if (!shaKmsKeyId) {
      throw new Error("SHA256_KMS_KEY_ARN environment variable is not set");
    }

    // Convert the message to bytes
    const messageBytes = Buffer.from(message, 'utf-8');

    // Create the command to generate the HMAC
    const command = new GenerateMacCommand({
      KeyId: shaKmsKeyId,
      Message: messageBytes,
      MacAlgorithm: SIGN_ALGORITHM,
    });

    // Use KMS to generate HMAC of the message
    const response = await kmsClient.send(command);

    // Base64 encode the signature for output
    const signature = Buffer.from(response.Mac as Uint8Array).toString('base64');

    // Create a result object to return the signature
    const resultDict = {
      "signature": signature
    };

    // Output the result as a JSON string
    console.log(JSON.stringify(resultDict));
  } catch (error) {
    // Handle and log any errors that occur
    console.error("Error:", error);
  }
}

// Execute the main function
main();
